File: 프로그래머스/2/42842. 카펫/카펫.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int brown, int yellow) {
    vector<int> answer(2);
    int total = brown + yellow;

    for(int w = 3; w <= total; w++) {
        int h = total / w;
        if (total % h == 0 && (w - 2) * (h - 2) == yellow) {
            answer[0] = max(w, h);
            answer[1] = min(w, h);
            break;
        }
    }
    return answer;
}

File: 프로그래머스/2/42842. 카펫/카펫.java
class Solution {
    public int[] solution(int brown, int yellow) {
        // w*h = brown + yellow
        // yellow = (w-2)(h-2) == wh - 2w - 2h + 4 
        // yellow = brown + yellow - 2w - 2h + 4
        // brown = 2w + 2h - 4
        int wh = brown + yellow;
        int w, h = 0;
        for (w = 1; w <= wh; w++) {
            if (wh % w == 0) h = wh / w;
            if ((w - 2) * (h - 2) == yellow) break;
        }
        return new int[]{Math.max(w, h), Math.min(w, h)};
    }
}